File: src/uid.zig
const std = @import("std");
const crypto = @import("std").crypto;
const mem = @import("std").mem;

pub const UID = struct {
    data: [120]u1,

    const Self = @This();

    pub inline fn eq(self: *const Self, other: *const Self) bool {
        return mem.eql(u1, &self.data, &other.data);
    }

    pub fn nil() Self {
        return .{ .data = [_]u1{0} ** 120 };
    }

    pub fn init() Self {
        var bits: [120]u1 = undefined;
        for (0..bits.len) |i| {
            bits[i] = if (crypto.random.boolean()) 1 else 0;
        }
        return .{ .data = bits };
    }

    inline fn bitsToByte(bits: *const [8]u1) u8 {
        return (@as(u8, @intCast(bits[0])) << 7) |
            (@as(u8, @intCast(bits[1])) << 6) |
            (@as(u8, @intCast(bits[2])) << 5) |
            (@as(u8, @intCast(bits[3])) << 4) |
            (@as(u8, @intCast(bits[4])) << 3) |
            (@as(u8, @intCast(bits[5])) << 2) |
            (@as(u8, @intCast(bits[6])) << 1) |
            (@as(u8, @intCast(bits[7])));
    }

    pub fn toBytes(self: *const Self) [120 / 8]u8 {
        var ret: [120 / 8]u8 = undefined;

        for (0..(120 / 8)) |i| {
            ret[i] = Self.bitsToByte(&[8]u1{
                self.data[i * 8],
                self.data[(i * 8) + 1],
                self.data[(i * 8) + 2],
                self.data[(i * 8) + 3],
                self.data[(i * 8) + 4],
                self.data[(i * 8) + 5],
                self.data[(i * 8) + 6],
                self.data[(i * 8) + 7],
            });
        }

        return ret;
    }

    inline fn bitsToNibble(bits: *const [4]u1) u4 {
        return (@as(u4, @intCast(bits[0])) << 3) | (@as(u4, @intCast(bits[1])) << 2) | (@as(u4, @intCast(bits[2])) << 1) | (@as(u4, @intCast(bits[3])));
    }

    fn bitsToNibbles(source: *const [120]u1, dest: *[30]u4) void {
        for (0..dest.*.len) |i| {
            const bits = [4]u1{ source[i * 4], source[i * 4 + 1], source[i * 4 + 2], source[i * 4 + 3] };
            dest[i] = Self.bitsToNibble(&bits);
        }
    }

    inline fn bitsTo5Bit(bits: *const [5]u1) u5 {
        return (@as(u5, @intCast(bits[0])) << 4) | (@as(u5, @intCast(bits[1])) << 3) | (@as(u5, @intCast(bits[2])) << 2) | (@as(u5, @intCast(bits[3])) << 1) | (@as(u5, @intCast(bits[4])));
    }

    fn bitsTo5Bits(source: *const [120]u1, dest: *[24]u5) void {
        for (0..dest.*.len) |i| {
            const bits = [5]u1{ source[i * 5], source[i * 5 + 1], source[i * 5 + 2], source[i * 5 + 3], source[i * 5 + 4] };
            dest[i] = Self.bitsTo5Bit(&bits);
        }
    }

    inline fn bitsTo6Bit(bits: *const [6]u1) u6 {
        return (@as(u6, @intCast(bits[0])) << 5) | (@as(u6, @intCast(bits[1])) << 4) | (@as(u6, @intCast(bits[2])) << 3) | (@as(u6, @intCast(bits[3])) << 2) | (@as(u6, @intCast(bits[4])) << 1) | (@as(u6, @intCast(bits[5])));
    }

    fn bitsTo6Bits(source: *const [120]u1, dest: *[20]u6) void {
        for (0..dest.*.len) |i| {
            const bits = [6]u1{ source[i * 6], source[i * 6 + 1], source[i * 6 + 2], source[i * 6 + 3], source[i * 6 + 4], source[i * 6 + 5] };
            dest[i] = Self.bitsTo6Bit(&bits);
        }
    }

    const CHARS16: []const u8 = "0123456789ABCDEF";
    const CHARS32: []const u8 = "0123456789ABCDEFGHIJKLMNOPQRSTUV";
    const CHARS64: []const u8 = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz-_";

    pub fn toString16(self: *const Self, buffer: *[30]u8) void {
        var out: [30]u4 = undefined;
        Self.bitsToNibbles(&self.data, &out);
        for (0..30) |i| {
            buffer.*[i] = CHARS16[out[i]];
        }
    }

    pub fn toString32(self: *const Self, buffer: *[24]u8) void {
        var out: [24]u5 = undefined;
        Self.bitsTo5Bits(&self.data, &out);
        for (0..24) |i| {
            buffer.*[i] = CHARS32[out[i]];
        }
    }

    pub fn toString64(self: *const Self, buffer: *[20]u8) void {
        var out: [20]u6 = undefined;
        Self.bitsTo6Bits(&self.data, &out);
        for (0..20) |i| {
            buffer.*[i] = CHARS64[out[i]];
        }
    }

    fn in(ch: u8, options: []const u8) struct { found: bool, index: usize } {
        for (options, 0..) |option, i| {
            if (ch == option) {
                return .{ .found = true, .index = i };
            }
        }
        return .{ .found = false, .index = 0 };
    }

    inline fn nibbleToBits(nibble: u4) [4]u1 {
        return .{
            @as(u1, @intCast((nibble >> 3) & 1)),
            @as(u1, @intCast((nibble >> 2) & 1)),
            @as(u1, @intCast((nibble >> 1) & 1)),
            @as(u1, @intCast(nibble & 1)),
        };
    }

    fn nibblesToBits(source: *const [30]u4, dest: *[120]u1) void {
        for (0..source.*.len) |i| {
            @memmove(dest.*[i * 4 .. (i * 4) + 4], &Self.nibbleToBits(source.*[i]));
        }
    }

    inline fn bit5ToBits(bits: u5) [5]u1 {
        return .{
            @as(u1, @intCast((bits >> 4) & 1)),
            @as(u1, @intCast((bits >> 3) & 1)),
            @as(u1, @intCast((bits >> 2) & 1)),
            @as(u1, @intCast((bits >> 1) & 1)),
            @as(u1, @intCast(bits & 1)),
        };
    }

    fn bits5ToBits(source: *const [24]u5, dest: *[120]u1) void {
        for (0..source.*.len) |i| {
            @memmove(dest.*[i * 5 .. (i * 5) + 5], &Self.bit5ToBits(source.*[i]));
        }
    }

    inline fn bit6ToBits(bits: u6) [6]u1 {
        return .{
            @as(u1, @intCast((bits >> 5) & 1)),
            @as(u1, @intCast((bits >> 4) & 1)),
            @as(u1, @intCast((bits >> 3) & 1)),
            @as(u1, @intCast((bits >> 2) & 1)),
            @as(u1, @intCast((bits >> 1) & 1)),
            @as(u1, @intCast(bits & 1)),
        };
    }

    fn bits6ToBits(source: *const [20]u6, dest: *[120]u1) void {
        for (0..source.*.len) |i| {
            @memmove(dest.*[i * 6 .. (i * 6) + 6], &Self.bit6ToBits(source.*[i]));
        }
    }

    pub const UIDError = error{
        UIDParseError,
    };

    pub fn fromString16(buffer: *const [30]u8) !Self {
        var data_parts: [30]u4 = undefined;
        for (0..30) |i| {
            const result = Self.in(buffer[i], CHARS16);
            if (result.found) {
                data_parts[i] = @intCast(result.index);
            } else {
                return UIDError.UIDParseError;
            }
        }
        var data: [120]u1 = undefined;
        Self.nibblesToBits(&data_parts, &data);
        return Self{ .data = data };
    }

    pub fn fromString32(buffer: *const [24]u8) !Self {
        var data_parts: [24]u5 = undefined;
        for (0..24) |i| {
            const result = Self.in(buffer[i], CHARS32);
            if (result.found) {
                data_parts[i] = @intCast(result.index);
            } else {
                return UIDError.UIDParseError;
            }
        }
        var data: [120]u1 = undefined;
        Self.bits5ToBits(&data_parts, &data);
        return Self{ .data = data };
    }

    pub fn fromString64(buffer: *const [20]u8) !Self {
        var data_parts: [20]u6 = undefined;
        for (0..20) |i| {
            const result = Self.in(buffer[i], CHARS64);
            if (result.found) {
                data_parts[i] = @intCast(result.index);
            } else {
                return UIDError.UIDParseError;
            }
        }
        var data: [120]u1 = undefined;
        Self.bits6ToBits(&data_parts, &data);
        return Self{ .data = data };
    }
};

test "UID Demo [not an actual test]" {
    const uid1 = UID.init();
    var buf16: [30]u8 = undefined;
    var buf32: [24]u8 = undefined;
    var buf64: [20]u8 = undefined;
    uid1.toString16(&buf16);
    uid1.toString32(&buf32);
    uid1.toString64(&buf64);
    std.debug.print("UID = {s}, {s}, {s}\n", .{ buf16, buf32, buf64 });
    const uid16 = try UID.fromString16(&buf16);
    const uid32 = try UID.fromString32(&buf32);
    const uid64 = try UID.fromString64(&buf64);
    try std.testing.expect(uid1.eq(&uid16));
    try std.testing.expect(uid1.eq(&uid32));
    try std.testing.expect(uid1.eq(&uid64));
}

File: src/uid.cpp
#include "uid.hpp"
#include <cstring>
#include <optional>
#include <random>
// #include "effolkronium/random.hpp"

namespace dracon {

    // using UIDRandom = effolkronium::basic_random_static<std::random_device>;

    UID::UID() { data.reset(); }
    UID::UID(const std::bitset<120> & data) : data{data} { }
    auto UID::Generate() -> UID {
        auto                        ret = UID();
        std::array<unsigned int, 4> rng;
        static auto                 rd = std::random_device{};
        for (std::size_t i = 0; i < rng.size(); i++) { rng.at(i) = rd(); }
        rngBufferToBits(rng, ret.data);
        return ret;
    }
    auto UID::Empty() -> UID { return UID(); }
    auto UID::ToBytes() -> std::array<std::uint8_t, 120 / 8> {
        std::array<std::uint8_t, 120 / 8> ret{};

        for (std::size_t i = 0; i < (120 / 8); i++) {
            std::bitset<8> bits{};
            bits[0] = data[i * 8];
            bits[1] = data[(i * 8) + 1];
            bits[2] = data[(i * 8) + 2];
            bits[3] = data[(i * 8) + 3];
            bits[4] = data[(i * 8) + 4];
            bits[5] = data[(i * 8) + 5];
            bits[6] = data[(i * 8) + 6];
            bits[7] = data[(i * 8) + 7];
            ret[i]  = UID::bitsToByte(bits);
        }
        return ret;
    }
    auto UID::operator==(const UID & rhs) const -> bool { return data == rhs.data; }
    auto UID::ToString16() const -> std::string {
        std::string buffer{};
        buffer.resize(30);
        std::array<std::uint8_t, 30> out;
        bitsToNibbles(data, out);
        for (std::size_t i = 0; i < 30; i++) { buffer.at(i) = CHARS16[out[i]]; }
        return buffer;
    }
    auto UID::ToString32() const -> std::string {
        std::string buffer{};
        buffer.resize(24);
        std::array<std::uint8_t, 24> out;
        bitsTo5Bits(data, out);
        for (std::size_t i = 0; i < 24; i++) { buffer.at(i) = CHARS32[out[i]]; }
        return buffer;
    }
    auto UID::ToString64() const -> std::string {
        std::string buffer{};
        buffer.resize(20);
        std::array<std::uint8_t, 20> out;
        bitsTo6Bits(data, out);
        for (std::size_t i = 0; i < 20; i++) { buffer.at(i) = CHARS64[out[i]]; }
        return buffer;
    }
    auto UID::FromString16(const std::string_view buffer) -> std::optional<UID> {
        std::array<std::uint8_t, 30> data_parts{};
        for (std::size_t i = 0; i < 30; i++) {
            const auto result = in(buffer.at(i), CHARS16);
            if (result.has_value()) {
                data_parts.at(i) = static_cast<std::uint8_t>(result.value());
            } else {
                return std::nullopt;
            }
        }
        std::bitset<120> data{};
        nibblesToBits(data_parts, data);
        return UID{data};
    }
    auto UID::FromString32(const std::string_view buffer) -> std::optional<UID> {
        std::array<std::uint8_t, 24> data_parts{};
        for (std::size_t i = 0; i < 24; i++) {
            const auto result = in(buffer.at(i), CHARS32);
            if (result.has_value()) {
                data_parts.at(i) = static_cast<std::uint8_t>(result.value());
            } else {
                return std::nullopt;
            }
        }
        std::bitset<120> data{};
        bits5ToBits(data_parts, data);
        return UID{data};
    }
    auto UID::FromString64(const std::string_view buffer) -> std::optional<UID> {
        std::array<std::uint8_t, 20> data_parts{};
        for (std::size_t i = 0; i < 20; i++) {
            const auto result = in(buffer.at(i), CHARS64);
            if (result.has_value()) {
                data_parts.at(i) = static_cast<std::uint8_t>(result.value());
            } else {
                return std::nullopt;
            }
        }
        std::bitset<120> data{};
        bits6ToBits(data_parts, data);
        return UID{data};
    }
    auto UID::bitsToByte(const std::bitset<8> & bits) -> std::uint8_t {
        return static_cast<std::uint8_t>(bits.to_ulong() & 0xFF);
    }
    auto UID::bitsToNibble(const std::bitset<4> & bits) -> std::uint8_t {
        return static_cast<std::uint8_t>(bits.to_ulong() & 0xF);
    }
    auto UID::bitsToNibbles(const std::bitset<120> & source, std::array<std::uint8_t, 30> & dest) -> void {
        for (std::size_t i = 0; i < dest.size(); i++) {
            std::bitset<4> bits{};
            bits[0]    = source[i * 4];
            bits[1]    = source[(i * 4) + 1];
            bits[2]    = source[(i * 4) + 2];
            bits[3]    = source[(i * 4) + 3];
            dest.at(i) = bitsToNibble(bits);
        }
    }
    auto UID::bitsTo5Bit(const std::bitset<5> & bits) -> std::uint8_t {
        return static_cast<std::uint8_t>(bits.to_ulong() & 0b11111);
    }
    auto UID::bitsTo5Bits(const std::bitset<120> & source, std::array<std::uint8_t, 24> & dest) -> void {
        for (std::size_t i = 0; i < dest.size(); i++) {
            std::bitset<5> bits{};
            bits[0]    = source[i * 5];
            bits[1]    = source[(i * 5) + 1];
            bits[2]    = source[(i * 5) + 2];
            bits[3]    = source[(i * 5) + 3];
            bits[4]    = source[(i * 5) + 4];
            dest.at(i) = bitsTo5Bit(bits);
        }
    }
    auto UID::bitsTo6Bit(const std::bitset<6> & bits) -> std::uint8_t {
        return static_cast<std::uint8_t>(bits.to_ulong() & 0b111111);
    }
    auto UID::bitsTo6Bits(const std::bitset<120> & source, std::array<std::uint8_t, 20> & dest) -> void {
        for (std::size_t i = 0; i < dest.size(); i++) {
            std::bitset<6> bits{};
            bits[0]    = source[i * 6];
            bits[1]    = source[(i * 6) + 1];
            bits[2]    = source[(i * 6) + 2];
            bits[3]    = source[(i * 6) + 3];
            bits[4]    = source[(i * 6) + 4];
            bits[5]    = source[(i * 6) + 5];
            dest.at(i) = bitsTo6Bit(bits);
        }
    }
    auto UID::in(char ch, const char * const options) -> std::optional<std::size_t> {
        for (std::size_t i = 0; i < std::strlen(options); i++) {
            if (ch == options[i]) { return i; }
        }
        return std::nullopt;
    }
    auto UID::nibbleToBits(std::uint8_t nibble) -> std::bitset<4> {
        std::bitset<4> bits;
        bits[0] = (nibble >> 3) & 1;
        bits[1] = (nibble >> 2) & 1;
        bits[2] = (nibble >> 1) & 1;
        bits[3] = nibble & 1;
        return bits;
    }
    auto UID::nibblesToBits(const std::array<std::uint8_t, 30> & source, std::bitset<120> & dest) -> void {
        for (std::size_t i = 0; i < source.size(); i++) {
            auto bits         = nibbleToBits(source.at(i));
            dest[i * 4]       = bits[0];
            dest[(i * 4) + 1] = bits[1];
            dest[(i * 4) + 2] = bits[2];
            dest[(i * 4) + 3] = bits[3];
        }
    }
    auto UID::bit5ToBits(std::uint8_t bits) -> std::bitset<5> {
        std::bitset<5> obits;
        obits[0] = (bits >> 4) & 1;
        obits[1] = (bits >> 3) & 1;
        obits[2] = (bits >> 2) & 1;
        obits[3] = (bits >> 1) & 1;
        obits[4] = bits & 1;
        return obits;
    }
    auto UID::bits5ToBits(const std::array<std::uint8_t, 24> & source, std::bitset<120> & dest) -> void {
        for (std::size_t i = 0; i < source.size(); i++) {
            auto bits         = bit5ToBits(source.at(i));
            dest[i * 5]       = bits[0];
            dest[(i * 5) + 1] = bits[1];
            dest[(i * 5) + 2] = bits[2];
            dest[(i * 5) + 3] = bits[3];
            dest[(i * 5) + 4] = bits[4];
        }
    }
    auto UID::bit6ToBits(std::uint8_t bits) -> std::bitset<6> {
        std::bitset<6> obits;
        obits[0] = (bits >> 5) & 1;
        obits[1] = (bits >> 4) & 1;
        obits[2] = (bits >> 3) & 1;
        obits[3] = (bits >> 2) & 1;
        obits[4] = (bits >> 1) & 1;
        obits[5] = bits & 1;
        return obits;
    }
    auto UID::bits6ToBits(const std::array<std::uint8_t, 20> & source, std::bitset<120> & dest) -> void {
        for (std::size_t i = 0; i < source.size(); i++) {
            auto bits         = bit6ToBits(source.at(i));
            dest[i * 6]       = bits[0];
            dest[(i * 6) + 1] = bits[1];
            dest[(i * 6) + 2] = bits[2];
            dest[(i * 6) + 3] = bits[3];
            dest[(i * 6) + 4] = bits[4];
            dest[(i * 6) + 5] = bits[5];
        }
    }

    auto UID::rngDataToBits(unsigned int data) -> std::bitset<32> {
        std::bitset<32> obits;
        for (std::size_t i = 0, off = 31; i < 32; i++, off--) { obits[i] = (data >> off) & 1; }
        return obits;
    }
    auto UID::rngBufferToBits(const std::array<unsigned int, 4> & source, std::bitset<120> & dest) -> void {
        for (std::size_t i = 0; i < source.size(); i++) {
            auto bits = rngDataToBits(source.at(i));
            for (std::size_t j = 0; j < 32; j++) {
                if (((i * 32) + j) >= 120) { break; }
                dest[(i * 32) + j] = bits[j];
            }
        }
    }

} // namespace dracon
